docs/design/data-protection-test.md

```diff
@@ -0,0 +1,219 @@
+
+This design addresses that need by measuring:
+
+- **Upload speed:** How quickly data is transferred to object storage.
+- **Snapshot performance:** How long a CSI VolumeSnapshot(s) take to become ReadyToUse.
+- **S3 compatibility:** Identifying the specific S3 vendor by inspecting HTTP response headers from the storage endpoint.
+
+## Goals
+
+- Upload data and compute the speed in Mbps.
+- Create a CSI VolumeSnapshot(s) and measure the time taken for it to become ready (reported as a duration).
+- Identify the S3-compatible vendor using an HTTP call that inspects response headers.
+- Expose all results—including upload speed, snapshot ready duration, and S3 vendor—in the CRD status.
+- Must gather will be updated th gather the DPT CR(s) for insights.
+
+
+## Non-Goals
+
+- This design will not create or modify BackupStorageLocation entries in OADP.
+- It will not implement download or latency tests, focusing solely on upload speed.
+- Scheduling of recurring tests is not supported in the initial version.
+- Integration via DPA CR
+
+## High-Level Design
+Components involved and their responsibilities:
+
+- **DataProtectionTest (DPT) CRD:**
+    - **Spec:**
+        - **backupLocation:** Contains Velero-based backup storage configuration.
+        - **uploadSpeedTestConfig:** Test parameters for the upload (file size, test timeout).
+        - **CSIVolumeSnapshotTestConfig:** Test parameters for the CSI VolumeSnapshot test (snapshot class, source PVC name, and importantly the PVC namespace, plus snapshot timeout).
+    - **Status:**
+        - **UploadTestStatus:** Groups upload speed (in Mbps), success flag, and error messages.
+        - **SnapshotTestStatus:** Groups the snapshot test results, reporting status and the duration taken for the snapshot to be ready.
+        - **S3Vendor:** Reports the detected S3 vendor string from vendor determination.
+- **DataProtectionTest Controller:**
+    - Monitors DataProtectionTest CRs.
+    - Extracts configuration from the backup location spec.
+    - Determines the S3 vendor via an HTTP HEAD call.
+    - Initializes the appropriate cloud provider using the CloudProvider interface.
+    - Executes the upload test and, if enabled, the CSI VolumeSnapshot test.
+    - Updates the CRD status with grouped results.
+- **CloudProvider Interface:**
+    - Defines an `UploadTest(ctx, config, bucket, fileSizeMB) (int64, error)` function.
+    - AWS-specific implementation (S3Provider) is provided using the AWS SDK.
+- **Vendor Determination Logic:**
+    - A helper function performs an HTTP HEAD call to the **s3Url** and inspects headers (especially the `Server` header) to determine the vendor (e.g., "AWS", "MinIO", etc.).
+
+## Detailed Design
+
+#### Proposed DataProtectionTest CRD:
+
+DataProtectionTest (DPT) CRD would look like:
+
+```yaml
+apiVersion: oadp.openshift.io/v1alpha1
+kind: DataProtectionTest
+metadata:
+  name: my-data-protection-test
+spec:
+  backupLocation:
```

Comment:
Would it make sense to also have `spec.bslName` -- validate that only one of `bslName` and `backupLocation` can be present -- that way if the BSL is already defined in the cluster, they don't need to copy all of that metadata over -- they can just reference the BSL by name.

Reply:
Hmm, yeah we can do that, optional name, but then it would be either the BSLConfig or the BSLName, not both.

Reply:
Done, added a note and updated DPT CRD.